File: jsrc/kernelprolog/fluents/SxxMachine/IntegerSource.java
package SxxMachine;

import SxxMachine.LongTerm;
import SxxMachine.Term;
import SxxMachine.Prog;
import SxxMachine.Source;

/**
 * creates a source of integers based on x=a*x+b formula
 */
public class IntegerSource extends Source {

	public IntegerSource(long fuel, long a, long x, long b, Prog p) {
		super(p);
		this.fuel = fuel;
		this.a = a;
		this.b = b;
		this.x = x;
	}

	private long fuel;

	private long a;

	private long b;

	private long x;

	public Term getElement() {
		if (fuel <= 0)
			return null;
		LongTerm R = Long(x);
		x = a * x + b;
		--fuel;
		return R;
	}

	public void stop() {
		fuel = 0;
	}

	public String pprint() {
		return "{(x->" + a + "*x+" + b + ")[" + fuel + "]=" + x + "}";
	}

}

File: src/kernelprolog/fluents/SxxMachine/IntegerSource.cpp
using namespace std;

#include "IntegerSource.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/LongTerm.h"
#include "../../main/SxxMachine/Builtins.h"

namespace SxxMachine
{

	IntegerSource::IntegerSource(long long fuel, long long a, long long x, long long b, Prog *p) : Source(p)
	{
		this->fuel = fuel;
		this->a = a;
		this->b = b;
		this->x = x;
	}

	Term *IntegerSource::getElement()
	{
		if (fuel <= 0)
		{
			return nullptr;
		}
		LongTerm *R = optional<long long>(x);
		x = a * x + b;
		--fuel;
		return R;
	}

	void IntegerSource::stop()
	{
		fuel = 0;
	}

	wstring IntegerSource::pprint()
	{
		return "{(x->" + to_string(a) + "*x+" + to_string(b) + ")[" + to_string(fuel) + "]=" + to_string(x) + "}";
	}
}
